### ISSUE TITLE
Poor TCP throughput over subnet router

### ISSUE THREAD

**Issue by nelhage** [tailscale/tailscale#8290]
### What is the issue?

I'm seeing consistently poor TCP throughput when connecting to a node through a subnet router. 

In particular, my throughput is often half or less of the minimal throughput of (client<->router, router<->destination). I expect the connection through the subnet router to be able to saturate the bottleneck link.

### Steps to reproduce

My setup: I have three nodes; we'll call them "client", "router", and "dest". 

- "client" and "router" are on the tailnet
- "router" advertises (incidentally, using 4via6) a route that includes "dest".
- "router" and "dest" are in the same cloud environment
- "client" is in a different cloud region

On "dest" and "router" both, I start a simple TCP server that serves random bytes:
```
dest$ while :; do nc -l -p 9999 < /dev/urandom ; done
```
```
router$ while :; do nc -l -p 4000 < /dev/urandom ; done
```
(I use different ports only to reduce the odds of confusing myself).

We can then crudely measure TCP throughput on three different legs, using `nc | pv`:

## dest -> router
This is a fast network within a cloud environment in a single region, so we expect good performance here. The tailnet isn't involved, just validating that this link is fast.
```
router$ nc DEST 9999 | pv -s 100M -Sterb >/dev/null
 100MiB 0:00:00 [ 314MiB/s]
```

## router -> client
`dest` and `client` are in two different cloud regions so we expect somewhat slower performance.

This connection goes through the tailnet (I have verified it's a direct connection, no DERP), but should not implicate the subnet route.
```
client$ nc router 4000 | pv -s 100M -Sterb >/dev/null
 100MiB 0:00:01 [57.1MiB/s]
```

## dest -> client
This is the problematic link. I expect throughput approximately equal to the minimum of the above two numbers, but instead I consistently see *half* of that or worse.
```
client$ nc DESTv6 9999 | pv -s 100M -Sterb >/dev/null
 100MiB 0:00:05 [19.5MiB/s]
client$ nc DESTv6 9999 | pv -s 100M -Sterb >/dev/null
 100MiB 0:00:05 [19.9MiB/s]
```

# Notes
I fairly consistently see (often much) worse throughput through the subnet router than to it; e.g. if I run the experiment using a laptop on wifi as "client", I see worse baseline throughput for "router -> client", but I still see a substantial drop *compared to that baseline* for "dest -> client". I've reproduced using another node in a cloud environment and shared those results here to try to cut WiFi variance out of the picture.

### OS

Linux

### OS version

Alpine Linux 3.16.5

### Tailscale version

1.40.1

### Bug report

BUG-fb0d544252b762ba7c36824ad2d817f452be368e186893b3a65b4d9c766853ae-20230606234937Z-f6c83a7d199278e8

**Comment by DentonGentry** [comment]
Around the time of the bugreport we see:
`2023-06-06 23:45:46.457243942 +0000 UTC: Accept: TCP{[fd7a:115c:a1e0:ab12:q:r:s:t]:55184 > [fd7a:115c:a1e0:b1a:0:w:x:y]:2222} 72 tcp non-syn`

The destination is [a 4via6 address](https://tailscale.com/kb/1201/4via6-subnets/). Is the nc for throughput testing connecting to a 4via6 destination address?

If so: the 4via6 path does use netstack as described in https://github.com/tailscale/tailscale/issues/8266. It will terminate the TCP connection on each side and make a new outgoing connection on the other, and there is substantial overhead in moving data around.

**Comment by nelhage** [comment]
Yeah, this is a 4via6 connection. At the time I filed this I was under the mistaken impression that subnet routers used netstack forwarding regardless of whether 4via6 was involved.

We are likely going to investigate moving off of 4via6 (potentially via internal renumbering) which will hopefully resolve our internal needs; I'd personally be fine if you want to interpret this ticket as a request for netstack performance work, or as a request for #8266 .

**Comment by DentonGentry** [comment]
There is a path to better performing 4via6 subnet routing, but it involves fairly involved packet filter work. An iptables NAT filter can rewrite within an existing IP header, it can't rewrite an IPv4 header to be an IPv6 header or vice-versa.

It is also different work that is needed to fix https://github.com/tailscale/tailscale/issues/8266, where we have to implement Windows and macOS and other OS APIs.

I'll repurpose this as a feature request to improve 4via6 routing performance.
